### PR TITLE
Deprecate auto-application of empty argument lists to parameterless functions

### PR DESCRIPTION
**Pull request by jared-barocsi**
Scala 3 now enforces a distinction between nullary parameter-less functions and nullary functions with empty argument lists. Previously, Scala 2 would auto-apply an empty parameter list onto any invocation of a nullary parameter-less function.

As such, in order to migrate to Scala 3, any nullary function that doesn't induce a side effect, e.g. `asUInt`, must not include a parameter list in its invocation. To deprecate the use of empty parameter lists for these functions, dummy versions of these functions are included which explicitly allow use of an empty parameter list (for example, a default dummy argument or a vararg).

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

code refactoring

#### API Impact

No effect 

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

Squash and merge

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.3.x`, [small] API extension: `3.4.x`, API modification or big change: `3.5.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

File: core/src/main/scala/chisel3/Bits.scala [hunk 1]
```diff
@@ -25,7 +25,9 @@ private[chisel3] sealed trait ToBoolable extends Element {
     *
     * @note The width must be known and equal to 1
     */
-  final def asBool(): Bool = macro SourceInfoWhiteboxTransform.noArg
+  final def asBool: Bool = macro SourceInfoWhiteboxTransform.noArg
+  @deprecated("Calling a nullary function with an empty argument list will be removed by official release", "chisel3")
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/Bits.scala:29]:
```suggestion
  @deprecated("Calling a nullary function with an empty argument list is not valid in Scala3. Use .asBool without parentheses", "3.5")
```

I think a good deprecation message should as clearly as possible what you should do if you get this deprecation message. And I thought the second argument is the version in which this was first deprecated?

File: core/src/main/scala/chisel3/When.scala [hunk 2]
```diff
@@ -111,7 +111,9 @@ final class WhenContext private[chisel3] (
   def otherwise(block: => Any)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Unit =
     new WhenContext(sourceInfo, None, block, firrtlDepth + 1, cond ++: altConds)
 
-  def active(): Boolean = scopeOpen
+  def active: Boolean = scopeOpen
+  @deprecated("Calling a nullary function with an empty argument list will be removed by official release", "chisel3")
+  def active(dummy: Int = 0): Boolean = active
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/When.scala:116]: why did you sometimes use this Int-with-default form and sometimes use the variable argument list?

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 3]
```diff
@@ -485,7 +485,7 @@ private[chisel3] object Builder extends LazyLogging {
   }
 
   // Returns the prefix stack at this moment
-  def getPrefix(): Prefix = chiselContext.get().prefixStack
+  def getPrefix: Prefix = chiselContext.get().prefixStack
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/internal/Builder.scala:488]: why does this one not need a deprecation version?

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 4]
```diff
@@ -572,7 +572,7 @@ private[chisel3] object Builder extends LazyLogging {
     dynamicContext.whenStack = s
   }
 
-  def currentWhen(): Option[WhenContext] = dynamicContext.whenStack.headOption
+  def currentWhen: Option[WhenContext] = dynamicContext.whenStack.headOption
```
Comment (mwachs5) [inline core/src/main/scala/chisel3/internal/Builder.scala:575]: why does this one not need a deprecation version?

File: src/main/scala/chisel3/util/Decoupled.scala [hunk 5]
```diff
@@ -37,7 +37,9 @@ object ReadyValidIO {
 
     /** Indicates if IO is both ready and valid
      */
-    def fire(): Bool = target.ready && target.valid
+    def fire: Bool = target.ready && target.valid
```
Comment (mwachs5) [inline src/main/scala/chisel3/util/Decoupled.scala:40]: I think there is something in the docs that use `fire`. Which makes me wonder if we need to update the mdoc as part of this PR

**Comment by davidbiancolin** [comment]
Inb4 "`asUInt` _does_ have a side effect".

**Comment by aswaterman** [comment]
> Inb4 "`asUInt` _does_ have a side effect".

trigger warning

File: core/src/main/scala/chisel3/Bits.scala (hunk 1 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/Bits.scala:29]: Yes you should always tell the user what to do. I don't think it's necessary to say "Use `.asBool` without parentheses", you could just say "Use form without parentheses" or something, this would make it easier to copy-paste the deprecation method throughout.

File: core/src/main/scala/chisel3/Bits.scala [hunk 6]
```diff
@@ -876,10 +905,10 @@ sealed class SInt private[chisel3] (width: Width) extends Bits(width) with Num[S
   /** @group SourceInfoTransformMacro */
   def do_=== (that: SInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bool = compop(sourceInfo, EqualOp, that)
 
-//  final def abs(): UInt = macro SourceInfoTransform.noArg
+//  final def abs(): UInt = macro SourceInfoTransform.noArgDummy
 
   def do_abs(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): SInt = {
-    Mux(this < 0.S, (-this), this)
+    Mux(this < 0.S, (this.unary_-(sourceInfo, compileOptions)), this)
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:911]: Why is this change necessary? Do you get deprecation warnings any time you do `-x` now??

File: core/src/main/scala/chisel3/internal/Builder.scala (hunk 3 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/Builder.scala:488]: Because it's [package] private, we can do whatever we want. That being said, this includes a `.get()` of the dynamic context so I'd say keep the `()`. 

```suggestion
  def getPrefix(): Prefix = chiselContext.get().prefixStack
```

File: core/src/main/scala/chisel3/internal/Builder.scala (hunk 4 above)
Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/Builder.scala:575]: Same package private thing, doesn't need to be deprecated.

**Comment by jackkoenig** [comment]
> Inb4 "`asUInt` _does_ have a side effect".

Go away.

In all seriousness though for anyone wondering though--while basically any Chisel API does have side effects, we go to great lengths to hide those side effects from the user such that Chisel APIs _appear to be side effect free_ and therefore will follow the Scala style of not having parentheses.

**Review by jackkoenig — CHANGES_REQUESTED** [review]
Please follow Megan's suggestion on the deprecation messages (ie. they need to tell the user what to do). As I said in the comment, I don't think they need to mention the name of the method, but do need to say something like "use form without parentheses" or "use form of method without parentheses" or something.

Please also make the dummy argument style consistent. Since varargs are the only style that work for macros, I'd apply that throughout.

Also, please leave a blank line between a method definition and the start of the following definition (including when it's a deprecation annotation).

File: core/src/main/scala/chisel3/Bits.scala [hunk 7]
```diff
@@ -222,7 +224,9 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
     * @return this $coll with each bit inverted
     * @group Bitwise
     */
-  final def unary_~ (): Bits = macro SourceInfoWhiteboxTransform.noArg
+  final def unary_~ : Bits = macro SourceInfoWhiteboxTransform.noArg
+  @deprecated("Calling a nullary function with an empty argument list will be removed by official release", "chisel3")
+  final def unary_~(dummy: Int*): Bits = macro SourceInfoWhiteboxTransform.noArgDummy
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:229]:
```suggestion
  final def unary_~ : Bits = macro SourceInfoWhiteboxTransform.noArg

  @deprecated("Calling a nullary function with an empty argument list will be removed by official release", "chisel3")
  final def unary_~(dummy: Int*): Bits = macro SourceInfoWhiteboxTransform.noArgDummy
```

File: core/src/main/scala/chisel3/Bits.scala (hunk 6 above)
Comment (jared-barocsi) [reply core/src/main/scala/chisel3/Bits.scala:911]: The compiler was complaining about ambiguity between this and the dummy form.

Comment (jackkoenig) [reply core/src/main/scala/chisel3/Bits.scala:911]: Right, so that ambiguity is a huge problem because it basically means users can't use `-` in a unary way any more on `SInts`. This should've been caught by the tests but apparently we never finished this particular test (https://github.com/chipsalliance/chisel3/blob/958904cb2f2f65d02b2ab3ec6d9ec2e06d04e482/src/test/scala/chiselTests/SIntOps.scala#L47).

Please make that test be active and we need to find a way to not have the ambiguity because obviously `-mySInt` needs to work. Do you know why this same issue doesn't show up for stuff like `unary_!` on `Bool`? I tried it out locally, confirmed this issue but was not able to see the same issue on `Bool` which is very weird.

Comment (mwachs5) [reply core/src/main/scala/chisel3/Bits.scala:922]: @jared-barocsi did you add the test to be active? Seems like we still have not gotten to the bottom of this question?

Comment (jared-barocsi) [reply core/src/main/scala/chisel3/Bits.scala]: @mwachs5 Yes, I just now fixed the ambiguity issue

File: core/src/main/scala/chisel3/Bits.scala [hunk 8]
```diff
@@ -304,10 +310,14 @@ sealed abstract class Bits(private[chisel3] val width: Width) extends Element wi
   def do_>> (that: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Bits
 
   /** Returns the contents of this wire as a [[scala.collection.Seq]] of [[Bool]]. */
-  final def toBools(): Seq[Bool] = macro SourceInfoTransform.noArg
+  final def toBools: Seq[Bool] = macro SourceInfoTransform.noArg
+  @deprecated("Calling this function with an empty argument list is invalid in Scala 3. Use the form without parentheses instead", "3.5")
+  final def toBools(dummy: Int*): Seq[Bool] = macro SourceInfoWhiteboxTransform.noArgDummy
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Bits.scala:315]: Interestingly, `.toBools` is supposed to be gone and I'm a bit confused why it's here... Definitely not the fault of this PR but weird.

File: src/main/scala/chisel3/util/Decoupled.scala (hunk 5 above)
Comment (jackkoenig) [reply src/main/scala/chisel3/util/Decoupled.scala:40]: I checked and there are no warnings in the mdoc coming from this PR

**Review by jackkoenig — CHANGES_REQUESTED** [review]
Every deprecation warning needs the version to be `"Chisel 3.5"`, not just `"3.5"` because the Scala compiler either doesn't know or at least doesn't report the name of the project that deprecation warnings come from.

File: src/main/scala/chisel3/util/Decoupled.scala (hunk 5 above)
Comment (mwachs5) [reply src/main/scala/chisel3/util/Decoupled.scala:40]: cool

File: core/src/main/scala/chisel3/Bits.scala (hunk 8 above)
Comment (mwachs5) [reply core/src/main/scala/chisel3/Bits.scala:315]: I think its still only a deprecation warning?

Comment (mwachs5) [reply core/src/main/scala/chisel3/Bits.scala:315]: EDIT: seems like it still exists for `Bits`: https://github.com/chipsalliance/chisel3/blob/4b7b771eeced366345779a75987ce552558a1c7e/core/src/main/scala/chisel3/Bits.scala#L307
Should we get rid of that?

Comment (mwachs5) [reply core/src/main/scala/chisel3/Bits.scala:315]: oh. This is the line you are talking about 😂 . Yeah lets get rid of this one in another PR

Comment (sequencer) [reply core/src/main/scala/chisel3/Bits.scala]: > Interestingly, .toBools is supposed to be gone and I'm a bit confused why it's here... Definitely not the fault of this PR but weird.

My bad, I think did didn't full remove it in https://github.com/chipsalliance/chisel3/pull/1730

I think these should be removed, I can create another PR, or @jared-barocsi can do it in this PR.

Comment (mwachs5) [reply core/src/main/scala/chisel3/Bits.scala]: @sequencer please create another PR (I opened #2143 to track), as it's not really related to this PR

**Review by jackkoenig — APPROVED** [review]

**Comment by chick** [comment]
@jackkoenig @sequencer I think I might have figured this out. Looks like the macro expansions of do_unary_- are not defined for FixedPoint and Interval, not sure how everything worked before here but I'm testing a fix now

**Comment by sequencer** [comment]
super fast!
